src/gas_metering/mod.rs

```diff
@@ -101,84 +105,129 @@ impl Rules for ConstantCostRules {
+/// [`mutable_global`] for details.
 ///
 /// This routine runs in time linear in the size of the input module.
 ///
 /// The function fails if the module contains any operation forbidden by gas rule set, returning
-/// the original module as an Err.
-pub fn inject<R: Rules>(
+/// the original module as an `Err`.
+pub fn inject<R: Rules, B: Backend>(
 	module: elements::Module,
+	backend: B,
 	rules: &R,
-	gas_module_name: &str,
 ) -> Result<elements::Module, elements::Module> {
-	// Injecting gas counting external
+	// Prepare module and return the gas function
+	let gas_meter = backend.gas_meter(&module, rules);
+
+	let import_count = module.import_count(elements::ImportCountType::Function) as u32;
+	let functions_space = module.functions_space() as u32;
+	let gas_global_idx = module.globals_space() as u32;
+
 	let mut mbuilder = builder::from_module(module);
-	let import_sig =
-		mbuilder.push_signature(builder::signature().with_param(ValueType::I64).build_sig());
-
-	mbuilder.push_import(
-		builder::import()
-			.module(gas_module_name)
-			.field("gas")
-			.external()
-			.func(import_sig)
-			.build(),
-	);
 
-	// back to plain module
-	let mut module = mbuilder.build();
+	// Calcutate the indexes
+	let (gas_func_idx, total_func) = match gas_meter {
+		GasMeter::External { module: gas_module, function } => {
+			// Inject the import of the gas function
+			let import_sig = mbuilder
+				.push_signature(builder::signature().with_param(ValueType::I64).build_sig());
+			mbuilder.push_import(
+				builder::import()
+					.module(gas_module)
+					.field(function)
+					.external()
+					.func(import_sig)
+					.build(),
+			);
 
-	// calculate actual function index of the imported definition
-	//    (subtract all imports that are NOT functions)
+			(import_count, functions_space + 1)
+		},
+		GasMeter::Internal { global, .. } => {
+			// Inject the gas counting global
+			mbuilder.push_global(
+				builder::global()
+					.with_type(ValueType::I64)
+					.mutable()
+					.init_expr(Instruction::I64Const(0))
+					.build(),
+			);
+			// Inject the export entry for the gas counting global
+			let ebuilder = builder::ExportBuilder::new();
+			let global_export = ebuilder
+				.field(global)
+				.with_internal(elements::Internal::Global(gas_global_idx))
+				.build();
+			mbuilder.push_export(global_export);
+
+			let func_idx = functions_space;
+			(func_idx, func_idx + 1)
+		},
+	};
+
+	// Build the module
+	let mut module = mbuilder.build();
 
-	let gas_func = module.import_count(elements::ImportCountType::Function) as u32 - 1;
-	let total_func = module.functions_space() as u32;
 	let mut need_grow_counter = false;
 	let mut error = false;
 
-	// Updating calling addresses (all calls to function index >= `gas_func` should be incremented)
+	let mut gas_fn_cost = 0u64;
+	if let GasMeter::Internal { global: _, function: _, cost, .. } = &gas_meter {
+		gas_fn_cost = *cost;
+	};
```

Comment:
Mutability makes reasoning harder.

```suggestion
	let gas_fn_cost = if let GasMeter::Internal { global: _, function: _, cost, .. } = &gas_meter {
		gas_fn_cost = *cost;
	} else {
	        0
	};
```